File: DSA/C++/Inorder Traversal on Binary Tree (Non recursive).cpp
// In this program we will perform inorder traversal on binary tree (non recursively) or we can say using stack

#include <iostream>
using namespace std;

class Node {
public:
  int val;
  Node* left;
  Node* right;
    Node (int data){
      val = data;
      left = NULL;
      right = NULL;
    }
};

Node* BuildTree(Node* &root){
  int data;
  cout<<"Enter data:- ";
  cin>>data;

  if (data==-1) return NULL;

  root = new Node(data);

  cout<<"Enter data in left of "<<data<<endl;
  root->left = BuildTree(root->left);

  cout<<"Enter data in right of "<<data<<endl;
  root->right = BuildTree(root->right);

  return root;

}

class Stack{
public:
  Node* st[100];
  int top = -1;

  void push(Node* &root){
    st[++top] = root;
  }

  Node* pop(){
    return st[top--];
  }
};

void Inorder(Node* root) {    
  Stack s;  
  Node* current = root;

  while (current != NULL || s.top != -1) {
    while (current != NULL) {
      s.push(current);
      current = current->left;
    }
    current = s.pop();
    cout << current->val << " ";
    current = current->right;
  }
}


int main() {
  Node* root = NULL;
  BuildTree(root);
  cout<<"Inorder :- ";
  Inorder(root);
}

File: Data Structures And Algorithm/Java/Inorder Traversal on Binary Tree (Non recursive).java
// In this program we will perform inorder traversal on binary tree (non recursively) or we can say using stack.

import java.util.Scanner;
import java.util.Stack;

class Node {
    int val;
    Node left, right;

    Node(int data) {
        val = data;
        left = null;
        right = null;
    }
}

public class InorderTraversalNonRecursive {

    static Node buildTree() {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter data : ");
        int data = scanner.nextInt();

        if (data == -1) {
            return null;
        }

        Node root = new Node(data);

        System.out.println("Enter data in left of " + data);
        root.left = buildTree();

        System.out.println("Enter data in right of " + data);
        root.right = buildTree();

        return root;
    }

    static void inorder(Node root) {
        if (root == null) {
            return;
        }

        Stack<Node> stack = new Stack<>();
        Node current = root;

        while (current != null || !stack.isEmpty()) {
            while (current != null) {
                stack.push(current);
                current = current.left;
            }

            current = stack.pop();
            System.out.print(current.val + " ");

            current = current.right;
        }
    }

    public static void main(String[] args) {
        Node root = buildTree();

        System.out.println("Inorder Traversal is:");
        inorder(root);
        System.out.println();
    }
}
